puppet/lib/puppet/reports/puppetdb.rb

```diff
@@ -159,8 +159,13 @@ def run_duration
   # @api private
   def event_to_hash(event)
     corrective_change = defined?(event.corrective_change) ? event.corrective_change : nil
+
+    # PUP-1542 sometimes causes status to be nil; fix it up here so we
+    # can store the rest of the report
+    status = event.status || 'failure'
```

Comment:
...and I think where we settled later was that the agent may eventually put a final "status = 'failure'" in the enclosing "ensure" block (presumably when nil).  Assuming ruby guarantees that will always fire, then nil should become impossible.